api/api/src/main/scala/uk/ac/wellcome/platform/api/QueryParams.scala

```diff
@@ -0,0 +1,238 @@
+      params.validationErrors match {
+        case Nil => provide(params)
+        case errs =>
+          reject(ValidationRejection(errs.mkString(", ")))
+            .toDirective[Tuple1[MultipleWorksParams]]
+      }
+    }
+
+  implicit val workTypeFilter: Decoder[WorkTypeFilter] =
+    decodeCommaSeperated.emap(strs => Right(WorkTypeFilter(strs)))
+
+  implicit val itemLocationTypeFilter: Decoder[ItemLocationTypeFilter] =
+    decodeCommaSeperated.emap(strs => Right(ItemLocationTypeFilter(strs)))
+
+  implicit val languageFilter: Decoder[LanguageFilter] =
+    decodeCommaSeperated.emap(strs => Right(LanguageFilter(strs)))
+
+  implicit val genreFilter: Decoder[GenreFilter] =
+    Decoder.decodeString.emap(str => Right(GenreFilter(str)))
+
+  implicit val subjectFilter: Decoder[SubjectFilter] =
+    Decoder.decodeString.emap(str => Right(SubjectFilter(str)))
+
+  implicit val aggregationsDecoder: Decoder[List[AggregationRequest]] =
+    decodeOneOfCommaSeperated(
+      "workType" -> AggregationRequest.WorkType,
+      "genres" -> AggregationRequest.Genre,
+      "production.dates" -> AggregationRequest.ProductionDate,
+      "subjects" -> AggregationRequest.Subject,
+      "language" -> AggregationRequest.Language,
+    )
+
+  implicit val sortDecoder: Decoder[List[SortRequest]] =
+    decodeOneOfCommaSeperated(
+      "production.dates" -> ProductionDateSortRequest
+    )
+
+  implicit val sortOrderDecoder: Decoder[SortingOrder] =
+    decodeOneOf(
+      "asc" -> SortingOrder.Ascending,
+      "desc" -> SortingOrder.Descending,
+    )
+}
+
+trait QueryParamsUtils extends Directives with TimeInstances {
+
+  implicit def unmarshaller[T](
+    implicit decoder: Decoder[T]): Unmarshaller[String, T] =
+    Unmarshaller.strict[String, T] { str =>
+      decoder.decodeJson(Json.fromString(str)) match {
+        case Left(err)    => throw new IllegalArgumentException(err.message)
+        case Right(value) => value
+      }
+    }
+
+  implicit val includesDecoder: Decoder[V2WorksIncludes] =
+    decodeOneOfCommaSeperated(
+      V2WorksIncludes.recognisedIncludes.zip(
+        V2WorksIncludes.recognisedIncludes
+      ): _*
+    ).emap(values => Right(V2WorksIncludes(values)))
+
+  implicit val decodeLocalDate: Decoder[LocalDate] =
+    decodeLocalDateDefault.withErrorMessage(
+      "Invalid date encoding. Expected YYYY-MM-DD"
+    )
+
+  implicit val decodeInt: Decoder[Int] =
+    Decoder.decodeInt.withErrorMessage("must be a valid Integer")
+
+  def decodeCommaSeperated: Decoder[List[String]] =
+    Decoder.decodeString.emap(str => Right(str.split(",").toList))
+
+  def decodeOneOf[T](values: (String, T)*): Decoder[T] =
+    Decoder.decodeString.emap { str =>
+      values.toMap
+        .get(str)
+        .map(Right(_))
+        .getOrElse(Left(invalidValuesMsg(List(str), values.map(_._1).toList)))
+    }
+
+  def decodeOneOfCommaSeperated[T](values: (String, T)*): Decoder[List[T]] =
```

Comment:
I'm not 100% sure I get the naming of this? Unless I've mixed it up is it more `decodeCommaSeparatedList`?

Reply:
I think I wanted to get across the fact that each value in the list can be one of a set number of values (there is a corresponding `decodeOneOf` which only decodes a single value rather than a list).

How about `decodeEnum` and `decodeCommaSeperatedEnumList` instead?

Reply:
Maybe let's keep it at this, and we can run through it in person as a separate PR.